# Notebook 2: NumPy Mathematical Operations

**Sprint 3: NumPy & Pandas for AI/ML Engineers**

---

## What are NumPy Mathematical Operations?

NumPy gives ready-made functions to **summarize an array** (sum, mean, min, max, percentile) and to do **linear algebra** (dot product, matrix multiplication, transpose, inverse). They run on the whole array at C speed, without Python loops.

**Analogy:** A teacher has a mark sheet. Instead of adding marks one by one with a pen, she presses the "total", "average" and "highest" buttons on a calculator. NumPy functions are those buttons.

**Real-world example:** A weather station has 365 daily temperatures. One line gives the yearly average, hottest day and the day it happened.

**Business example:** A retail chain has monthly sales of 3 products. `sum(axis=0)` gives the total per product, `sum(axis=1)` gives the total per month, and `argmax` tells which product sold the most.

**AI/ML example:**
- Statistics (mean, std) are used for scaling features.
- **Dot product** is the core of a neuron: `output = inputs . weights + bias`.
- **Matrix multiplication** runs a whole neural network layer for all samples at once.
- **Transpose, inverse** are used in the Linear Regression formula (Sprint 2).

**Topics covered:** Sum, Mean, Median, Variance, Standard Deviation, Min, Max, Argmin, Argmax, Percentile, Dot Product, Matrix Multiplication, Transpose, Basic Linear Algebra.

In [1]:
import numpy as np

data = np.array([4, 8, 6, 5, 3, 9, 10])
print("data:", data)

data: [ 4  8  6  5  3  9 10]


## 1. Sum, Mean and Median

| Function | Meaning | Formula |
|---|---|---|
| `np.sum(a)` | Total of all elements | x1 + x2 + ... + xn |
| `np.mean(a)` | Average | sum / n |
| `np.median(a)` | Middle value of the sorted data | middle element |

**Numerical example:** Data = 4, 8, 6, 5, 3, 9, 10. Sum = 45, n = 7, Mean = 45 / 7 = **6.43**. Sorted = 3, 4, 5, 6, 8, 9, 10, the middle (4th) value is **6**, so Median = 6.

**Business example:** Average daily sales, and the median salary (not affected by a few very high salaries).

**AI/ML example:** Mean and median are used for filling missing values, and the mean is used in standardization and in loss functions like MSE.

**Axis:** For 2D arrays, `axis=0` gives one result **per column** (down the rows), and `axis=1` gives one result **per row** (across the columns).

In [2]:
print("Sum   :", np.sum(data))
print("Mean  :", round(np.mean(data), 2))
print("Median:", np.median(data))

# 2D array: axis
m = np.array([[1, 2, 3],
              [4, 5, 6]])
print("\nMatrix:\n", m)
print("Total sum        :", m.sum())
print("Sum axis=0 (column):", m.sum(axis=0))
print("Sum axis=1 (row)   :", m.sum(axis=1))
print("Mean axis=0        :", m.mean(axis=0))
print("Mean axis=1        :", m.mean(axis=1))

# Median is not affected by an outlier
salaries = np.array([25000, 28000, 30000, 32000, 200000])
print("\nMean salary  :", salaries.mean())
print("Median salary:", np.median(salaries))

# Business: monthly sales (4 months x 3 products)
sales = np.array([[120, 150, 130],
                  [200, 180, 210],
                  [90, 110, 100],
                  [300, 280, 320]])
print("\nTotal per product:", sales.sum(axis=0))
print("Total per month  :", sales.sum(axis=1))

Sum   : 45
Mean  : 6.43
Median: 6.0

Matrix:
 [[1 2 3]
 [4 5 6]]
Total sum        : 21
Sum axis=0 (column): [5 7 9]
Sum axis=1 (row)   : [ 6 15]
Mean axis=0        : [2.5 3.5 4.5]
Mean axis=1        : [2. 5.]

Mean salary  : 63000.0
Median salary: 30000.0

Total per product: [710 720 760]
Total per month  : [400 590 300 900]


**Output Explanation:**
- Sum is 45, mean is 6.43 and median is 6 for our data.
- For the 2 x 3 matrix, `axis=0` gives `[5, 7, 9]` (one value per column) and `axis=1` gives `[6, 15]` (one value per row). `axis=0` means "collapse the rows".
- The salary mean (63000) is pulled up by the single 200000, while the median (30000) stays typical. So the median is better for skewed data.
- In the sales table, `axis=0` gives the total of each product across months, and `axis=1` gives the total of each month across products.

## 2. Variance and Standard Deviation

| Function | Meaning |
|---|---|
| `np.var(a)` | Average of squared distances from the mean |
| `np.std(a)` | Square root of the variance (same unit as data) |

**Formula:**

$$\sigma^2 = \frac{\sum (x_i - \mu)^2}{N} \qquad s^2 = \frac{\sum (x_i - \bar{x})^2}{n - 1}$$

**Important:** NumPy uses `ddof=0` (population, divide by N) **by default**. For a sample, use `ddof=1` (divide by n - 1). Pandas uses `ddof=1` by default, so the same data can give different answers in NumPy and Pandas.

**Numerical example:** For our data, sum of squared differences = 41.71. Population variance = 41.71 / 7 = **5.96**. Sample variance = 41.71 / 6 = **6.95**. Standard deviation = **2.44** (population) and **2.64** (sample).

**Business example:** Two stores both average 100 sales per day. The one with the lower standard deviation is more predictable, so stock planning is easier.

**AI/ML example:** Standardization divides by the standard deviation. Features with zero variance are removed. PCA keeps the directions of highest variance.

In [3]:
print("Population variance:", round(np.var(data), 3))
print("Sample variance    :", round(np.var(data, ddof=1), 3))
print("Population std     :", round(np.std(data), 3))
print("Sample std         :", round(np.std(data, ddof=1), 3))

# Manual check
mean = data.mean()
manual_var = np.sum((data - mean) ** 2) / len(data)
print("\nManual population variance:", round(manual_var, 3))

# Same mean, different spread
steady = np.array([48, 49, 50, 51, 52])
risky = np.array([10, 30, 50, 70, 90])
print("\nMean (steady, risky):", steady.mean(), risky.mean())
print("Std  (steady, risky):", round(steady.std(), 2), round(risky.std(), 2))

# Column-wise standard deviation
print("\nStd of each product:", sales.std(axis=0).round(2))

Population variance: 5.959
Sample variance    : 6.952
Population std     : 2.441
Sample std         : 2.637

Manual population variance: 5.959

Mean (steady, risky): 50.0 50.0
Std  (steady, risky): 1.41 28.28

Std of each product: [81.36 62.85 85.15]


**Output Explanation:**
- Population variance is 5.959 and sample variance is 6.952. The sample value is bigger because we divide by n - 1 instead of n.
- Our manual calculation matches `np.var`, which confirms the formula.
- Both arrays have the same mean (50), but the std is about 1.41 for `steady` and 28.28 for `risky`, so the mean alone does not tell us about spread.
- `sales.std(axis=0)` gives the spread of each product across the 4 months.

## 3. Minimum, Maximum, Argmin and Argmax

| Function | Returns |
|---|---|
| `np.min(a)` | The smallest **value** |
| `np.max(a)` | The largest **value** |
| `np.argmin(a)` | The **index** (position) of the smallest value |
| `np.argmax(a)` | The **index** (position) of the largest value |
| `np.ptp(a)` | Range (max - min) |

**Analogy:** `max` tells you the highest mark (98). `argmax` tells you **who** got it (student number 4).

**Numerical example:** For 4, 8, 6, 5, 3, 9, 10: min = 3 at index 4, max = 10 at index 6.

**Business example:** Highest-selling month and which month it was. The best and worst performing branch.

**AI/ML example:** **`argmax` is used to get the predicted class**: a classifier gives probabilities `[0.1, 0.7, 0.2]`, and `argmax` returns class 1. It is also used to find the nearest neighbour (`argmin` of distances), and the best hyperparameter (`argmax` of scores).

In [4]:
print("Min   :", data.min(), "| index:", data.argmin())
print("Max   :", data.max(), "| index:", data.argmax())
print("Range :", np.ptp(data))

# 2D with axis
print("\nBest month for each product (row index):", sales.argmax(axis=0))
print("Best product in each month (col index) :", sales.argmax(axis=1))
print("Overall max in sales:", sales.max(), "| flat index:", sales.argmax())
print("Position as (row, col):", np.unravel_index(sales.argmax(), sales.shape))

# AI/ML: class prediction from probabilities
probs = np.array([[0.10, 0.70, 0.20],
                  [0.80, 0.15, 0.05],
                  [0.25, 0.25, 0.50]])
classes = np.array(["cat", "dog", "bird"])
pred_idx = probs.argmax(axis=1)
print("\nPredicted class index:", pred_idx)
print("Predicted class names:", classes[pred_idx])

# Nearest neighbour: smallest distance
distances = np.array([4.2, 1.3, 7.8, 2.5])
print("\nNearest neighbour index:", distances.argmin())


Min   : 3 | index: 4
Max   : 10 | index: 6
Range : 7

Best month for each product (row index): [3 3 3]
Best product in each month (col index) : [1 2 1 2]
Overall max in sales: 320 | flat index: 11
Position as (row, col): (np.int64(3), np.int64(2))

Predicted class index: [1 0 2]
Predicted class names: ['dog' 'cat' 'bird']

Nearest neighbour index: 1


**Output Explanation:**
- `min` and `max` give the values 3 and 10, while `argmin` and `argmax` give their **positions** (4 and 6).
- For a 2D array `argmax` without axis returns a position in the **flattened** array, so `np.unravel_index` converts it back to (row, column).
- `argmax(axis=1)` on the probability table picks the most probable class for each sample: `[1, 0, 2]`, which is dog, cat, bird.
- The smallest distance 1.3 is at index 1, so index 1 is the nearest neighbour.

## 4. Percentile

`np.percentile(a, q)` gives the value below which **q%** of the data falls. `np.quantile(a, q)` is the same but q is between 0 and 1.

| Percentile | Meaning |
|---|---|
| 25th (Q1) | 25% of values are below it |
| 50th | The median |
| 75th (Q3) | 75% of values are below it |

NumPy uses **linear interpolation** by default when the position falls between two values.

**Numerical example:** Sorted data = 3, 4, 5, 6, 8, 9, 10 (n = 7). Position = (n - 1) x q.
- 25th: position 1.5, between 4 and 5, so **4.5**
- 50th: position 3, so **6**
- 75th: position 4.5, between 8 and 9, so **8.5**

**Business example:** A delivery app tracks the 95th percentile of delivery time to promise "95% of orders arrive within X minutes".

**AI/ML example:** IQR outlier detection (Q3 - Q1), `RobustScaler`, clipping extreme values, and monitoring model latency (P95, P99).

In [5]:
print("25th percentile:", np.percentile(data, 25))
print("50th percentile:", np.percentile(data, 50))
print("75th percentile:", np.percentile(data, 75))
print("90th percentile:", np.percentile(data, 90))
print("Several at once:", np.percentile(data, [25, 50, 75]))
print("quantile(0.5)  :", np.quantile(data, 0.5))

# IQR outlier detection
values = np.array([12, 15, 17, 18, 20, 22, 25, 27, 30, 35, 90])
q1, q3 = np.percentile(values, [25, 75])
iqr = q3 - q1
lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
print("\nQ1:", q1, "| Q3:", q3, "| IQR:", iqr)
print("Fences:", lower, upper, "| Outliers:", values[(values < lower) | (values > upper)])

# Clip (cap) values at the 5th and 95th percentile
low, high = np.percentile(values, [5, 95])
print("\nClipped:", np.clip(values, low, high).round(1))

25th percentile: 4.5
50th percentile: 6.0
75th percentile: 8.5
90th percentile: 9.4
Several at once: [4.5 6.  8.5]
quantile(0.5)  : 6.0

Q1: 17.5 | Q3: 28.5 | IQR: 11.0
Fences: 1.0 45.0 | Outliers: [90]

Clipped: [13.5 15.  17.  18.  20.  22.  25.  27.  30.  35.  62.5]


**Output Explanation:**
- The 25th, 50th and 75th percentiles are 4.5, 6 and 8.5, matching our hand calculation. The 90th percentile is 9.4.
- `np.percentile(data, [25, 50, 75])` returns many percentiles in one call.
- For the second dataset, Q1 = 17.5, Q3 = 28.5, IQR = 11, upper fence = 45, so **90 is the outlier** (same as Sprint 2).
- `np.clip` replaces values beyond the 5th and 95th percentile with those limits, which keeps all rows but reduces the effect of extremes.

## 5. Dot Product

The dot product of two vectors multiplies the elements at the **same position** and **adds** the results. The answer is a **single number**.

**Formula:**

$$a \cdot b = \sum_{i=1}^{n} a_i b_i = a_1 b_1 + a_2 b_2 + \dots + a_n b_n$$

**Numerical example:** a = (1, 2, 3), b = (4, 5, 6)
a . b = 1x4 + 2x5 + 3x6 = 4 + 10 + 18 = **32**

**Real-world / business example:** Quantities bought (5, 2, 10) and prices (100, 250, 80). The dot product 5x100 + 2x250 + 10x80 = **1800** is the total bill.

**AI/ML example:**
- A neuron computes `inputs . weights + bias`.
- **Cosine similarity** between two vectors (documents, users) uses the dot product divided by the lengths.
- A linear model's prediction for one sample is `x . w + b`.

$$\cos(a, b) = \frac{a \cdot b}{\lVert a \rVert \, \lVert b \rVert}$$

**Note:** Dot product is **not** `a * b` (that is element-wise and gives an array, not one number).

In [6]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

print("Element-wise a * b:", a * b)
print("Dot product       :", np.dot(a, b))
print("Using @           :", a @ b)
print("Manual sum(a * b) :", np.sum(a * b))

# Business: total bill
qty = np.array([5, 2, 10])
price = np.array([100, 250, 80])
print("\nTotal bill:", qty @ price)

# AI/ML: one neuron
inputs = np.array([0.5, 0.8, 0.2])
weights = np.array([0.4, 0.3, 0.9])
bias = 0.1
z = inputs @ weights + bias
print("\nNeuron z      :", round(z, 3))
print("Neuron output :", round(1 / (1 + np.exp(-z)), 3), "(sigmoid)")

# AI/ML: cosine similarity
def cosine(u, v):
    return u @ v / (np.linalg.norm(u) * np.linalg.norm(v))

print("\nCosine (same direction) :", round(cosine(np.array([1, 2]), np.array([2, 4])), 3))
print("Cosine (similar tastes) :", round(cosine(np.array([5, 4]), np.array([4, 5])), 3))
print("Cosine (perpendicular)  :", round(cosine(np.array([1, 0]), np.array([0, 1])), 3))

Element-wise a * b: [ 4 10 18]
Dot product       : 32
Using @           : 32
Manual sum(a * b) : 32

Total bill: 1800

Neuron z      : 0.72
Neuron output : 0.673 (sigmoid)

Cosine (same direction) : 1.0
Cosine (similar tastes) : 0.976
Cosine (perpendicular)  : 0.0


**Output Explanation:**
- `a * b` gives `[4, 10, 18]` (element-wise), and the dot product adds them to **32**.
- The bill 1800 is a dot product of quantity and price.
- The neuron computes z = 0.5x0.4 + 0.8x0.3 + 0.2x0.9 + 0.1 = **0.72**, and the sigmoid converts it into a value between 0 and 1 (about 0.673).
- Cosine similarity is **1** for vectors in the same direction, about **0.976** for similar tastes, and **0** for perpendicular vectors (nothing in common).

## 6. Matrix Multiplication

Matrix multiplication combines **rows of A with columns of B**. Each result cell is a **dot product** of one row and one column.

**Rule:** For A of shape (m, n) and B of shape (n, p), the **inner dimensions must match** (n = n), and the result has shape **(m, p)**.

**Formula:**

$$C_{ij} = \sum_{k} A_{ik} B_{kj}$$

**Numerical example:**

A = [[1, 2], [3, 4]], B = [[5, 6], [7, 8]]
- C[0,0] = 1x5 + 2x7 = 19
- C[0,1] = 1x6 + 2x8 = 22
- C[1,0] = 3x5 + 4x7 = 43
- C[1,1] = 3x6 + 4x8 = 50

So A @ B = [[19, 22], [43, 50]]. Element-wise `A * B` would give [[5, 12], [21, 32]], which is a different thing.

**Business example:** A table of 3 stores x 2 products (units sold) multiplied by a 2 x 1 column of prices gives the revenue of each store.

**AI/ML example:** A neural network layer for the **whole dataset at once** is `outputs = X @ W + b`, where X is (samples, features) and W is (features, neurons). Linear Regression predictions for all rows are `X @ w`.

**Remember:** In general, `A @ B` is **not** equal to `B @ A`.

In [7]:
A = np.array([[1, 2],
              [3, 4]])
B = np.array([[5, 6],
              [7, 8]])

print("A @ B:\n", A @ B)
print("np.matmul:\n", np.matmul(A, B))
print("A * B (element-wise):\n", A * B)
print("B @ A:\n", B @ A)

# Shape rule
X = np.ones((3, 2))
Y = np.ones((2, 4))
print("\n(3,2) @ (2,4) ->", (X @ Y).shape)
try:
    Y @ X
except ValueError as e:
    print("(2,4) @ (3,2) -> Error:", str(e)[:60])

# Business: revenue of each store
units = np.array([[10, 5],
                  [20, 8],
                  [15, 12]])        # 3 stores x 2 products
prices = np.array([[100],
                   [250]])          # 2 products x 1
print("\nRevenue per store:\n", units @ prices)

# AI/ML: layer for 3 samples, 2 features, 2 neurons
X = np.array([[1.0, 2.0],
              [3.0, 4.0],
              [5.0, 6.0]])
W = np.array([[0.5, 1.0],
              [1.0, 0.5]])
b = np.array([1.0, 0.0])
print("\nLayer output (X @ W + b):\n", X @ W + b)

# Linear regression prediction
w = np.array([0.5, 1.0])
print("Predictions (X @ w + 1):", X @ w + 1)

A @ B:
 [[19 22]
 [43 50]]
np.matmul:
 [[19 22]
 [43 50]]
A * B (element-wise):
 [[ 5 12]
 [21 32]]
B @ A:
 [[23 34]
 [31 46]]

(3,2) @ (2,4) -> (3, 4)
(2,4) @ (3,2) -> Error: matmul: Input operand 1 has a mismatch in its core dimension

Revenue per store:
 [[2250]
 [4000]
 [4500]]

Layer output (X @ W + b):
 [[3.5 2. ]
 [6.5 5. ]
 [9.5 8. ]]
Predictions (X @ w + 1): [3.5 6.5 9.5]


**Output Explanation:**
- `A @ B` is `[[19, 22], [43, 50]]`, same as our hand calculation. `A * B` is a different result (element-wise), and `B @ A` is also different, so **order matters**.
- A (3, 2) matrix times a (2, 4) matrix gives (3, 4). The reverse order fails because the inner dimensions (4 and 3) do not match.
- Revenue per store: 10x100 + 5x250 = 2250, 20x100 + 8x250 = 4000, 15x100 + 12x250 = 4500.
- The layer output is computed for all 3 samples in a single line, and the bias `b` is added by **broadcasting** (Notebook 1).
- Linear regression predictions: `[3.5, 6.5, 9.5]`.

## 7. Transpose

The transpose **swaps rows and columns**. Element (i, j) moves to (j, i). A (2, 3) array becomes (3, 2).

**Formula:** (A^T)[i, j] = A[j, i]

**Numerical example:** A = [[1, 2, 3], [4, 5, 6]] gives A^T = [[1, 4], [2, 5], [3, 6]].

**Useful rules:** (A^T)^T = A, and (A @ B)^T = B^T @ A^T.

**Business example:** A table has stores as rows and months as columns. The transpose makes months the rows, which is easier to read for some reports.

**AI/ML example:** The **normal equation** of Linear Regression uses `X.T @ X`. The **covariance matrix** is `X.T @ X / (n - 1)` for centered data. `X.T` is also used in backpropagation.

**Note:** `.T` returns a **view** (no data is copied). For a 1D array, `.T` does nothing.

In [8]:
A = np.array([[1, 2, 3],
              [4, 5, 6]])
print("A shape:", A.shape)
print("A.T:\n", A.T, "| shape:", A.T.shape)
print("np.transpose(A) same?:", np.array_equal(A.T, np.transpose(A)))
print("(A.T).T equals A?    :", np.array_equal(A.T.T, A))

# 1D array: transpose does nothing
v = np.array([1, 2, 3])
print("\n1D v.T shape:", v.T.shape, "-> use reshape to make it a column:", v.reshape(-1, 1).shape)

# X.T @ X gives a square matrix (features x features)
X = np.array([[1, 2], [3, 4], [5, 6]])
print("\nX.T @ X:\n", X.T @ X)

# Covariance matrix by hand
hours = np.array([2, 4, 6, 8, 10], dtype=float)
marks = np.array([50, 60, 65, 80, 90], dtype=float)
D = np.column_stack([hours, marks])
D_centered = D - D.mean(axis=0)
cov_manual = D_centered.T @ D_centered / (len(D) - 1)
print("\nManual covariance matrix:\n", cov_manual)
print("np.cov check:\n", np.cov(hours, marks))

A shape: (2, 3)
A.T:
 [[1 4]
 [2 5]
 [3 6]] | shape: (3, 2)
np.transpose(A) same?: True
(A.T).T equals A?    : True

1D v.T shape: (3,) -> use reshape to make it a column: (3, 1)

X.T @ X:
 [[35 44]
 [44 56]]

Manual covariance matrix:
 [[ 10.  50.]
 [ 50. 255.]]
np.cov check:
 [[ 10.  50.]
 [ 50. 255.]]


**Output Explanation:**
- A has shape (2, 3), and `A.T` has shape (3, 2). Transposing twice returns the original.
- A 1D array stays the same after `.T`. To get a column, we use `reshape(-1, 1)`.
- `X.T @ X` for a (3, 2) matrix is a (2, 2) square matrix. This is the building block of the Linear Regression formula.
- Our manual covariance matrix (centered data, `D.T @ D / (n - 1)`) matches `np.cov`: variance of hours = 10, covariance = 50, variance of marks = 255.

## 8. Basic Linear Algebra Operations (`np.linalg`)

| Function | What it gives | Use |
|---|---|---|
| `np.linalg.det(A)` | Determinant | 0 means the matrix cannot be inverted |
| `np.linalg.inv(A)` | Inverse matrix (A @ A^-1 = I) | Solving equations |
| `np.linalg.solve(A, b)` | Solution x of A x = b | Preferred over `inv` (faster, more accurate) |
| `np.linalg.norm(v)` | Length (magnitude) of a vector | Distances, normalization |
| `np.linalg.matrix_rank(A)` | Number of independent rows/columns | Detecting redundant features |
| `np.linalg.eig(A)` | Eigenvalues and eigenvectors | PCA |
| `np.trace(A)` | Sum of diagonal elements | |
| `np.eye(n)` | Identity matrix | |

**Numerical example (solve equations):**
2x + y = 5
x + 3y = 10
In matrix form: A = [[2, 1], [1, 3]], b = [5, 10]. The solution is **x = 1, y = 3** (check: 2 + 3 = 5 and 1 + 9 = 10).

For A = [[1, 2], [3, 4]]: determinant = 1x4 - 2x3 = **-2**, and the inverse is [[-2, 1], [1.5, -0.5]].

**Business example:** Finding the price of two products from two total bills: a set of linear equations, solved with `solve`.

**AI/ML example:**
- **Linear Regression (normal equation):** w = (X^T X)^-1 X^T y.
- **PCA:** eigenvectors of the covariance matrix are the principal components, and eigenvalues are the variances.
- **Multicollinearity:** if the matrix rank is lower than the number of features, some features are duplicates (Sprint 2).
- Vector norm is used for L2 normalization, distances and regularization.

In [9]:
A = np.array([[1, 2],
              [3, 4]], dtype=float)

print("Determinant:", round(np.linalg.det(A), 2))
inv = np.linalg.inv(A)
print("Inverse:\n", inv)
print("A @ inverse (identity):\n", (A @ inv).round(2))
print("Trace:", np.trace(A), "| Rank:", np.linalg.matrix_rank(A))
print("Identity:\n", np.eye(2))

# Solve equations: 2x + y = 5, x + 3y = 10
coef = np.array([[2, 1],
                 [1, 3]], dtype=float)
rhs = np.array([5, 10], dtype=float)
sol = np.linalg.solve(coef, rhs)
print("\nSolution x, y:", sol)
print("Check coef @ sol:", coef @ sol)

# Vector norm
print("\nNorm of [3, 4]:", np.linalg.norm(np.array([3, 4])))

# Singular matrix (rows are multiples of each other)
S = np.array([[1, 2],
              [2, 4]], dtype=float)
print("\nSingular matrix det:", round(np.linalg.det(S), 2), "| rank:", np.linalg.matrix_rank(S))
try:
    np.linalg.inv(S)
except np.linalg.LinAlgError as e:
    print("Error:", e)

# Eigenvalues
E = np.array([[2, 0],
              [0, 3]], dtype=float)
vals, vecs = np.linalg.eig(E)
print("\nEigenvalues:", vals)

# Linear Regression with the normal equation: marks = b0 + b1 * hours
hours = np.array([2, 4, 6, 8, 10], dtype=float)
marks = np.array([50, 60, 65, 80, 90], dtype=float)
X = np.column_stack([np.ones(len(hours)), hours])     # add intercept column
w = np.linalg.solve(X.T @ X, X.T @ marks)
print("\nIntercept, slope:", w.round(2))
print("Predicted marks for 7 hours:", round(w[0] + w[1] * 7, 2))

Determinant: -2.0
Inverse:
 [[-2.   1. ]
 [ 1.5 -0.5]]
A @ inverse (identity):
 [[1. 0.]
 [0. 1.]]
Trace: 5.0 | Rank: 2
Identity:
 [[1. 0.]
 [0. 1.]]

Solution x, y: [1. 3.]
Check coef @ sol: [ 5. 10.]

Norm of [3, 4]: 5.0

Singular matrix det: 0.0 | rank: 1
Error: Singular matrix

Eigenvalues: [2. 3.]

Intercept, slope: [39.  5.]
Predicted marks for 7 hours: 74.0


**Output Explanation:**
- The determinant of A is -2 (not zero), so the inverse exists. `A @ inverse` gives the **identity matrix**, which proves the inverse is correct.
- `solve` gives x = 1 and y = 3, and the check `coef @ sol` returns the original `[5, 10]`.
- The norm of (3, 4) is sqrt(9 + 16) = 5.
- The matrix `S` has determinant 0 and rank 1 (the second row is twice the first), so it is **singular** and `inv` raises `LinAlgError`. In ML this is what happens with perfectly duplicated features (multicollinearity).
- Eigenvalues of a diagonal matrix are its diagonal values, 2 and 3.
- The normal equation gives intercept = **39** and slope = **5**, so marks = 39 + 5 x hours, and 7 hours predicts **74** marks. This matches the Linear Regression hand calculation from Sprint 2, now done using only NumPy.

In [10]:
with_nan = np.array([10.0, 20.0, np.nan, 40.0])
print("mean with NaN   :", np.mean(with_nan))
print("nanmean         :", np.nanmean(with_nan))
print("nansum / nanmax :", np.nansum(with_nan), np.nanmax(with_nan))
print("nanstd          :", round(np.nanstd(with_nan), 3))

# keepdims keeps the shape so broadcasting still works
m = np.array([[1.0, 2.0, 3.0],
              [4.0, 5.0, 6.0]])
row_mean = m.mean(axis=1, keepdims=True)
print("\nrow_mean shape:", row_mean.shape)
print("Centered by row:\n", m - row_mean)

# cumulative functions
nums = np.array([1, 2, 3, 4, 5])
print("\ncumsum :", np.cumsum(nums))
print("cumprod:", np.cumprod(nums))
print("diff   :", np.diff(np.array([100, 120, 150, 140])))

mean with NaN   : nan
nanmean         : 23.333333333333332
nansum / nanmax : 70.0 40.0
nanstd          : 12.472

row_mean shape: (2, 1)
Centered by row:
 [[-1.  0.  1.]
 [-1.  0.  1.]]

cumsum : [ 1  3  6 10 15]
cumprod: [  1   2   6  24 120]
diff   : [ 20  30 -10]


**Output Explanation:**
- If the array has a `NaN`, normal `mean` returns `nan`. The `nan` versions (`nanmean`, `nansum`, `nanmax`, `nanstd`) **ignore** missing values and give 23.33 for the mean.
- `keepdims=True` keeps the result as shape `(2, 1)` instead of `(2,)`, so `m - row_mean` broadcasts correctly.
- `cumsum` gives running totals `[1, 3, 6, 10, 15]`, `cumprod` gives `[1, 2, 6, 24, 120]`, and `diff` gives the change between consecutive values `[20, 30, -10]` (useful for day-to-day changes in sales).

## Best Practices

- Use NumPy functions (`np.sum`, `np.mean`) instead of Python `sum()` and loops on arrays.
- Always specify **`axis`** for 2D data, and check the result shape.
- Use **`ddof=1`** for sample variance and standard deviation.
- Use **`@`** for matrix multiplication and `*` only for element-wise.
- Use `np.linalg.solve(A, b)` instead of `np.linalg.inv(A) @ b`.
- Use the `nan` functions (`nanmean`, ...) when data has missing values.
- Check the shapes before matrix multiplication (inner dimensions must match).
- Use `keepdims=True` when the result will be used for broadcasting.

## Common Mistakes

| Mistake | Problem | Fix |
|---|---|---|
| Using `*` instead of `@` | Element-wise result, not matrix multiplication | Use `@` or `np.dot` |
| Forgetting `ddof=1` | Population value instead of sample value | `np.std(a, ddof=1)` |
| Mixing up `axis=0` and `axis=1` | Result of the wrong direction | axis=0 down the rows (per column), axis=1 across (per row) |
| Confusing `max` and `argmax` | Value instead of position | `max` is the value, `argmax` is the index |
| Calling `inv` on a singular matrix | `LinAlgError` | Check `det` and `rank`, or use `lstsq` / `pinv` |
| Mean of data with NaN | Result is `nan` | Use `np.nanmean` |
| Matrix multiplication with wrong shapes | `ValueError` | Inner dimensions must match |
| Assuming `A @ B == B @ A` | Wrong results | Matrix multiplication is not commutative |
| Comparing NumPy and Pandas std directly | Different values (ddof) | Set `ddof` the same in both |

## Advantages
- One-line, fast statistics and linear algebra on millions of values
- Same functions work for 1D, 2D and higher dimensions through `axis`
- Linear algebra tools (`np.linalg`) are built in
- Forms the base of ML algorithms: regression, PCA, neural networks

## Limitations
- Works on numeric data only, and the full array must fit in memory
- No labels (no column names), only positions
- `np.linalg.inv` can be numerically unstable for nearly singular matrices
- Runs on CPU, so very large matrix multiplications need GPU libraries (PyTorch, TensorFlow)
- Percentile and quantile methods may differ slightly from other tools